File: modules/webtoon-reader/ios/WebtoonReaderModule.swift
import ExpoModulesCore
import UIKit

public class WebtoonReaderModule: Module {
  public func definition() -> ModuleDefinition {
    Name("WebtoonReader")

    View(WebtoonReaderView.self) {
      Prop("data") { (view: WebtoonReaderView, data: [[String: Any]]) in
        view.updateData(data: data)
      }

      Prop("magnifierConfig") { (view: WebtoonReaderView, config: [String: Any]?) in
        view.updateMagnifierConfig(config: config)
      }

      Events(
        "onEndReached",
        "onChapterChanged",
        "onSingleTap",
        "onPageChanged",
        "onScrollBegin",
        "onLoadingStateChanged",
        "onImageError",
        "onRetryRequested"
      )

      AsyncFunction("scrollToIndex") { (view: WebtoonReaderView, chapterId: String, index: Int) in
        view.scrollToIndex(chapterId: chapterId, index: index)
      }

      AsyncFunction("getCurrentPosition") { (view: WebtoonReaderView) -> [String: Any] in
        return view.getCurrentPosition()
      }

      AsyncFunction("setZoomScale") { (view: WebtoonReaderView, scale: Double) in
        view.setZoomScale(scale: CGFloat(scale))
      }

      AsyncFunction("resetZoom") { (view: WebtoonReaderView) in
        view.resetZoom()
      }

      AsyncFunction("resetSession") { (view: WebtoonReaderView) in
        view.resetSession()
      }
    }
  }
}

File: modules/webtoon-reader/src/WebtoonReaderModule.ts
import { NativeModule, requireNativeModule } from 'expo';

declare class WebtoonReaderModule extends NativeModule<Record<string, never>> {
  scrollToIndex(viewRef: any, chapterId: string, index: number): Promise<void>;
  getCurrentPosition(viewRef: any): Promise<{ chapterId: string; pageIndex: number }>;
  setZoomScale(viewRef: any, scale: number): Promise<void>;
  resetZoom(viewRef: any): Promise<void>;
  resetSession(viewRef: any): Promise<void>;
}

// This call loads the native module object from the JSI.
export default requireNativeModule<WebtoonReaderModule>('WebtoonReader');
